### PR TITLE
feat(dart_frog_cli): start daemon implementation

### PR DESCRIPTION
**Pull request by renancaraujo**
## Status

**READY**

## Description

Adds the first methods and events to the daemon and the communication structure via stdio. 

We will need some e2e after this is merged. 

This is part of https://github.com/VeryGoodOpenSource/dart_frog/issues/705

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: packages/dart_frog_cli/lib/src/daemon/daemon_server.dart [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+  /// Whether the daemon has exited.
+  bool get isCompleted => _exitCodeCompleter.isCompleted;
+
+  /// The version of the Dart Frog daemon protocol.
+  String get version => daemonVersion;
+
+  /// Adds a [domain] to the daemon.
+  ///
+  /// Visible for testing purposes only.
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/daemon_server.dart:53]: Nit: already explicit with the annotation.

```suggestion
```

File: packages/dart_frog_cli/lib/src/daemon/daemon_server.dart [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+  void _sendMessage(DaemonMessage message) {
+    if (isCompleted) {
+      return;
+    }
+    _connection.outputSink.add(message);
+  }
+
+  /// Sends an [event] to the client.
+  void sendEvent(DaemonEvent event) {
+    _sendMessage(event);
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/daemon_server.dart:108]: What's the reason/idea behind this wrapper function?

File: packages/dart_frog_cli/lib/src/daemon/domain.dart [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        error: {
+          'message': 'Method not found: ${request.method}',
+        },
+      );
+    }
+    return handler(request);
+  }
+
+  /// Disposes of this domain.
+  Future<void> dispose();
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/domain.dart:48]: Nit: Should we also use the mustBeOverriden annotation here?

File: packages/dart_frog_cli/lib/src/daemon/domain.dart [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    );
+  }
+
+  /// Kills the daemon.
+  Future<DaemonResponse> _kill(DaemonRequest request) async {
+    daemon.kill(ExitCode.success).ignore();
+    return DaemonResponse.success(
+      id: request.id,
+      result: const {
+        'message': 'Hogarth. You stay, I go. No following.',
```
Comment (alestiago) [inline packages/dart_frog_cli/lib/src/daemon/domain.dart:102]: 💙🤖

**Review by alestiago — COMMENTED** [review]
Overall, looks well written and documented.

I haven't yet reached the test files. Need to check those to on my side.

I know we've agreed to let the daemon be within the CLI. Now, that it requires a version and so on, do you think it would be more maintainable as a standalone package that the CLI would use?

File: packages/dart_frog_cli/lib/src/daemon/connection.dart [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+/// - [DaemonStdioConnection], a connection that hooks into the stdio.
+abstract interface class DaemonConnection {
+  /// A stream of [DaemonMessage]s that are received from the client.
+  Stream<DaemonMessage> get inputStream;
+
+  /// A sink of [DaemonMessage]s that are to be sent to the client.
+  StreamSink<DaemonMessage> get outputSink;
+
+  /// Closes the connection and free resources.
+  Future<void> dispose();
```
[the remark this answers is not in the document]
Comment (renancaraujo) [reply packages/dart_frog_cli/lib/src/daemon/connection.dart:23]: As an abstract method, subclasses are already forced to override.

File: packages/dart_frog_cli/lib/src/daemon/daemon_server.dart (hunk 1 above)
Comment (renancaraujo) [reply packages/dart_frog_cli/lib/src/daemon/daemon_server.dart:53]: The annotation looks quite obscure on dart doc generated pages. That's why I always like to add this to the dardoc.

File: packages/dart_frog_cli/lib/src/daemon/daemon_server.dart (hunk 2 above)
Comment (renancaraujo) [reply packages/dart_frog_cli/lib/src/daemon/daemon_server.dart:108]: Daemon domains won't send requests and responses spontaneously.

Comment (renancaraujo) [reply packages/dart_frog_cli/lib/src/daemon/daemon_server.dart:108]: This only accepts events.

**Comment by renancaraujo** [comment]
>  **now** that it requires a version and so on
It always had a version, including on the POC. 

The server could well live outside the CLI, but as it depends on the logic of the bricks, dart_frog_gen, and the devserver runner, all of this would have to be extracted from the CLI as well. 

Maybe we can pack all of this on dart_frog_gen, but that problem is for the future.

File: packages/dart_frog_cli/lib/src/daemon/connection.dart [hunk 6]
```diff
@@ -0,0 +1,119 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:dart_frog_cli/src/daemon/daemon.dart';
+import 'package:meta/meta.dart';
+
+/// {@template daemon_connection}
+/// A connection of the daemon that hooks into an mean of communication with
+/// clients.
```
[the remark this answers is not in the document]
Comment (alestiago) [reply packages/dart_frog_cli/lib/src/daemon/connection.dart]: I like it picasso 🖌️

**Review by alestiago — APPROVED** [review]
lgtm

File: packages/dart_frog_cli/lib/src/daemon/connection.dart [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                  ),
+                );
+              case FormatException(message: _):
+                outputSink.add(
+                  const DaemonEvent(
+                    domain: DaemonDomain.name,
+                    event: 'protocolError',
+                    params: {'message': 'Not a valid JSON'},
+                  ),
+                );
```
Comment (wolfenrain) [inline packages/dart_frog_cli/lib/src/daemon/connection.dart:70]: Should we add a default case to prevent new errors from never being handled?

**Review by wolfenrain — APPROVED** [review]

**Review by alestiago — DISMISSED** [review]

File: packages/dart_frog_cli/lib/src/daemon/connection.dart [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  final StreamSink<List<int>> _stdout;
+  final Stream<List<int>> _stdin;
+
+  late final StreamController<DaemonMessage> _inputStreamController =
+      StreamController<DaemonMessage>();
+
+  late final StreamController<DaemonMessage> _outputStreamController =
+      StreamController<DaemonMessage>();
```
Comment (erickzanardo) [inline packages/dart_frog_cli/lib/src/daemon/connection.dart:93]:
```suggestion
  late final _inputStreamController = StreamController<DaemonMessage>();

  late final _outputStreamController = StreamController<DaemonMessage>();
```

**Review by erickzanardo — APPROVED** [review]

**Pull request merged**